packages/swingset-liveslots/src/virtualObjectManager.js

```diff
@@ -586,273 +661,282 @@ export function makeVirtualObjectManager(
 
-      function ensureState() {
-        if (innerSelf.rawState) {
-          cache.refresh(innerSelf);
-        } else {
-          innerSelf = cache.lookup(innerSelf.baseRef, true);
-        }
-      }
+    let isStateProperty = _prop => undefined;
+    /** @type {(value: any, prop: string) => void} */
+    let checkStateProperty = (_value, _prop) => undefined;
+    if (stateShape) {
+      isStateProperty = prop =>
+        hasOwnPropertyOf(stateShape, prop) ||
+        assert.fail(
+          X`State must only have fields described by stateShape: ${q(
+            ownKeys(stateShape),
+          )}`,
+        );
+      checkStateProperty = (value, prop) => {
+        isStateProperty(prop);
+        mustMatch(value, stateShape[prop]);
+      };
+    }
 
+    // The dataCache holds both unserialized and still-serialized
+    // (capdata) contents of the virtual-object state record.
+    // dataCache[baseRef] -> { capdatas, valueMap }
+    // valueCD=capdatas[prop], value=valueMap.get(prop)
+    /** @type { import('./cache.js').Cache<{ capdatas: any, valueMap: Map<string, any> }>} */
+    const dataCache = makeDataCache(syscall, unserialize);
+    allCaches.push(dataCache);
+
+    // Behavior functions will receive a 'state' object that provides
+    // access to their virtualized data, with getters and setters
+    // backed by the vatstore DB. When those functions are invoked and
+    // we miss in contextCache, we'll call makeState() and
+    // makeContext(). The makeState() call might read from the
+    // vatstore DB if we miss in dataCache.
+
+    // We sample dataCache.get() once each time:
+    // * makeState() is called, which happens the first time in each crank that
+    //   a method is invoked (and the prototype does getContext)
+    // * when state.prop is read, invoking the getter
+    // * when state.prop is written, invoking the setter
+    // This will cause a syscall.vatstoreGet only once per crank.
+
+    function makeState(baseRef) {
       const state = {};
-      if (!initializing) {
-        ensureState();
-      }
-      for (const prop of Object.getOwnPropertyNames(innerSelf.rawState)) {
+      for (const prop of Object.getOwnPropertyNames(
+        dataCache.get(baseRef).capdatas,
+      )) {
+        isStateProperty(prop);
         Object.defineProperty(state, prop, {
           get: () => {
-            ensureState();
-            return unserializeSlot(innerSelf.rawState[prop], prop);
+            const value = dataCache.get(baseRef).valueMap.get(prop);
+            checkStateProperty(value, prop);
+            return value;
           },
           set: value => {
-            ensureState();
-            const before = innerSelf.rawState[prop];
-            const after = serializeSlot(value, prop);
-            assertAcceptableSyscallCapdataSize([after]);
+            checkStateProperty(value, prop);
+            const capdata = serialize(value);
+            assertAcceptableSyscallCapdataSize([capdata]);
```

Comment:
This is a property-level check, but my understanding from https://github.com/Agoric/agoric-sdk/blob/master/packages/swingset-liveslots/src/vatstore-usage.md#virtual-object-kinds is that full state is stored as a single serialization. But it also seems acceptable to temporarily exceed full-state limits, e.g. `state.big2 = state.big1; state.big1 = undefined;`. Should there be some explanatory comments about when and how it is appropriate to check this, particularly explaining why it does or does not belong before any particular syscall?

Reply:
Huh, you're right, we're appying the check to one property at a time, but the syscall limit is actually going to be applied to the joint serialization of all the properties together (plus whatever overhead the `JSON.stringify` will add). The `capdatas` parameter name in liveslots.js where this is implemented suggests that we meant it to look at them in aggregate, but we're only passing in a single value.

On one hand, if the one property is so large that it trips the aggregate limit all by itself, it'd be nice to tell them which property caused the problem.

I think it's not hard to pass the whole capdatas list in, but I need to think about how to avoid creating a couple of temporary arrays. I'll add a TODO comment describing the incomplete check, and file a ticket to remind us to check it more thoroughly before we get to MN-3. (This check exists to convert a vat-fatal oversized syscall into a mere thrown Error, and it isn't immediately clear to me that this is useful, although I'm pretty sure I'm the one who advocated for it in the first place).

Reply:
https://github.com/Agoric/agoric-sdk/issues/7308